src/raqm.c

```diff
@@ -728,6 +767,432 @@ _raqm_hb_dir (raqm_t *rq, FriBidiLevel level)
+		/*WJ*/ { SS, XX, XX, SS, SS, SS, XX, XX, XX, SS, SS, SS, SS, SS, SS, SS, SS, SS, SS, XX, cc, XX, SS, SS, SS, SS, SS }, // WJ word joiner
+		/*H2*/ { oo, XX, XX, SS, SS, SS, XX, XX, XX, oo, SS, oo, oo, oo, SS, SS, SS, oo, oo, XX, cc, XX, oo, oo, oo, SS, SS }, // Hangul 2 Jamo syllable
+		/*H3*/ { oo, XX, XX, SS, SS, SS, XX, XX, XX, oo, SS, oo, oo, oo, SS, SS, SS, oo, oo, XX, cc, XX, oo, oo, oo, oo, SS }, // Hangul 3 Jamo syllable
+		/*JL*/ { oo, XX, XX, SS, SS, SS, XX, XX, XX, oo, SS, oo, oo, oo, SS, SS, SS, oo, oo, XX, cc, XX, SS, SS, SS, SS, oo }, // Jamo Leading Consonant
+		/*JV*/ { oo, XX, XX, SS, SS, SS, XX, XX, XX, oo, SS, oo, oo, oo, SS, SS, SS, oo, oo, XX, cc, XX, oo, oo, oo, SS, SS }, // Jamo Vowel
+		/*JT*/ { oo, XX, XX, SS, SS, SS, XX, XX, XX, oo, SS, oo, oo, oo, SS, SS, SS, oo, oo, XX, cc, XX, oo, oo, oo, oo, SS }, // Jamo Trailing Consonant
+	};
+
+	break_actions = malloc (sizeof (enum break_action) * length);
+	break_here = malloc (sizeof (bool) * length);
+
+	current_class = ucdn_get_resolved_linebreak_class(rq->text[0]);
+	next_class    = ucdn_get_resolved_linebreak_class(rq->text[1]);
+
+	// handle case where input starts with an LF
+	if (current_class == LF)
+		current_class = BK;
+
+	// treat NL like BK
+	if (current_class == NL)
+		current_class = BK;
+
+	// treat SP at start of input as if it followed WJ
+	if (current_class == SP)
+		current_class = WJ;
+
+	// loop over all pairs in the string up to a hard break or CRLF pair
+	for (size_t i = 1; (i < length) && (current_class != BK) && (current_class != CR || next_class == LF); i++)
+	{
+		next_class = ucdn_get_resolved_linebreak_class(rq->text[i]);
+
+		// handle spaces explicitly
+		if (next_class == SP) {
+                        break_actions[i-1]  = PROHIBITED_BREAK;   // apply rule LB 7: � SP
+			continue;
+		}
+
+		if (next_class == BK || next_class == NL || next_class == LF) {
+                        break_actions[i-1]  = PROHIBITED_BREAK;
+			current_class = BK;
+			continue;
+		}
+
+		if (next_class == CR)
+		{
+                        break_actions[i-1]  = PROHIBITED_BREAK;
+			current_class = CR;
+			continue;
+		}
+
+		// lookup pair table information in brkPairs[before, after];
+		current_action = break_pairs[current_class][next_class];
+		break_actions[i-1] = current_action;
+
+                if (current_action == INDIRECT_BREAK)		// resolve indirect break
+		{
+			if (ucdn_get_resolved_linebreak_class(rq->text[i-1]) == SP)                    // if context is A SP * B
+                                break_actions[i-1] = INDIRECT_BREAK;             //       break opportunity
+			else                                        // else
+                                break_actions[i-1] = PROHIBITED_BREAK;           //       no break opportunity
+		}
+
+                else if (current_action == COMBINING_PROHIBITED_BREAK)		// this is the case OP SP* CM
+		{
```

Comment:
Lines should be 80-chars long, you can place the comment above the code. Also don’t use C++ comments.